tfhe/src/shortint/engine/public_side.rs

```diff
@@ -179,35 +207,111 @@ impl ShortintEngine {
             / (public_key.parameters.message_modulus().0 * public_key.parameters.carry_modulus().0)
                 as u64;
 
-        //The input is reduced modulus the message_modulus
-        let m = message % message_modulus.0 as u64;
+        let encoded: Vec<_> = messages
+            .into_iter()
+            .map(move |message| {
+                //The input is reduced modulus the message_modulus
+                let m = message % message_modulus.0 as u64;
 
-        let shifted_message = m * delta;
-        // encode the message
-        let plain = Plaintext(shifted_message);
+                let shifted_message = m * delta;
+                // encode the message
+                Plaintext(shifted_message)
+            })
+            .collect();
 
         // This allocates the required ct
-        let mut encrypted_ct = LweCiphertext::new(
-            0u64,
-            public_key.lwe_public_key.lwe_size(),
-            public_key.lwe_public_key.ciphertext_modulus(),
+        let mut encrypted_ct = vec![
+            LweCiphertext::new(
+                0u64,
+                public_key.lwe_public_key.lwe_size(),
+                public_key.lwe_public_key.ciphertext_modulus(),
+            );
+            encoded.len()
+        ];
+
+        encrypt_lwe_ciphertext_iterator_with_seeded_public_key(
+            &public_key.lwe_public_key,
+            encrypted_ct.iter_mut().map(|lwe| lwe.as_mut_view()),
+            encoded,
+            &mut self.secret_generator,
         );
 
-        encrypt_lwe_ciphertext_with_seeded_public_key(
+        encrypted_ct
+            .into_iter()
+            .map(|lwe| {
+                Ciphertext::new(
+                    lwe,
+                    Degree::new(message_modulus.0 - 1),
+                    NoiseLevel::NOMINAL,
+                    message_modulus,
+                    CarryModulus(carry_modulus),
+                    public_key.pbs_order,
+                )
+            })
+            .collect()
+    }
+
+    pub(crate) fn encrypt_with_many_message_moduli_and_compressed_public_key(
```

Comment:
Same thing here, maybe `encrypt_many_with_message_modulus_and_compressed_public_key` could use this function internally with a single modulus

Reply:
as discussed the many does not apply to the same thing, renamed the engine methods a bit to make it less confusing